File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/238-product-of-array-except-self.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func productExceptSelf(nums []int) []int {
	// TODO: implement
	return nil
}

func intsEqual(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func runTests() {
	assert(intsEqual(productExceptSelf([]int{1, 2, 3, 4}), []int{24, 12, 8, 6}), "test 1")
	assert(intsEqual(productExceptSelf([]int{-1, 1, 0, -3, 3}), []int{0, 0, 9, 0, 0}), "test 2")
	assert(intsEqual(productExceptSelf([]int{1, 1}), []int{1, 1}), "test 3")
	assert(intsEqual(productExceptSelf([]int{2, 3}), []int{3, 2}), "test 4")
	assert(intsEqual(productExceptSelf([]int{1, 0}), []int{0, 1}), "test 5")
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/238-product-of-array-except-self.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function productExceptSelf(nums: number[]): number[] {
    // TODO: implement
    return [];
}

function _runTests(): void {
    assert(JSON.stringify(productExceptSelf([1, 2, 3, 4])) === JSON.stringify([24, 12, 8, 6]));
    assert(JSON.stringify(productExceptSelf([-1, 1, 0, -3, 3])) === JSON.stringify([0, 0, 9, 0, 0]));
    assert(JSON.stringify(productExceptSelf([1, 1])) === JSON.stringify([1, 1]));
    assert(JSON.stringify(productExceptSelf([2, 3])) === JSON.stringify([3, 2]));
    assert(JSON.stringify(productExceptSelf([1, 0])) === JSON.stringify([0, 1]));
    // perf
    const t0 = performance.now();
    productExceptSelf(Array.from({ length: 10000 }, (_, i) => i + 1));
    console.log(`perf productExceptSelf n=10000: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
